src/wallet/spend.cpp

```diff
@@ -186,6 +186,23 @@ void AvailableCoins(const CWallet& wallet, std::vector<COutput>& vCoins, const C
                 continue;
             }
 
+            OutputType output_type{wallet.m_default_address_type};
+            std::vector<std::vector<uint8_t>> dummy;
+            TxoutType type{Solver(wtx.tx->vout[i].scriptPubKey, dummy)};
+            if (type == TxoutType::WITNESS_V1_TAPROOT) {
+                output_type = OutputType::BECH32M;
+            } else if (type == TxoutType::WITNESS_V0_KEYHASH) {
+                output_type = OutputType::BECH32;
+            } else if (type == TxoutType::SCRIPTHASH) {
+                output_type = OutputType::P2SH_SEGWIT;
```

Comment:
In b32b52c3e7360ecb30e1a46dfa06669bc4fcd4b3 "wallet: avoid mixing ouput types in coin selection"

This is not guaranteed to be true. We could have a P2SH multisig and that would be `LEGACY`.

Reply:
is there a cleaner way to go map a `COutput` to an `OutputType`? adding lots of if/else checks to go from a `TxoutType` to an `OutputType` seems clunky to me

Reply:
Is there any need to map to output type? Why not use `TxoutType` directly?

Reply:
> Is there any need to map to output type? Why not use `TxoutType` directly?

Yes, `P2SH_SEGWIT` is distinct from other P2SH which would be legacy.

Reply:
> This is not guaranteed to be true. We could have a P2SH multisig and that would be `LEGACY`.

so if a `P2SH_SEWGIT` and a `P2SH` multisig were spent together, when the UTXO is later spent it would reveal that some inputs were legacy multisig and others p2sh-segwit? 

As an alternative, would it be better to map `WITNESS_V0_SCRIPTHASH` to `P2SH` and map `SCRIPTHASH` and `PUBKEYHASH` to `LEGACY`?

Reply:
By the time we get determining the output type, we already know that the output either belongs to the wallet, or is watch only. Thus we can do additional solver invocations for scripthash with the redeemScript in order to figure out the true type.

For watch only outputs, we may not be able to do that because not all of the scripts will necessarily be available. However we could still use solving data provided in `coinControl` to try to determine that.

Reply:
thanks for pointing me in the right direction! I'll take a stab at implementing this

Reply:
Now also considering the reedem script when a UTXO is `P2SH`